telemetry/tracks/class-tracks-event.php

```diff
@@ -0,0 +1,336 @@
+<?php
+/**
+ * Telemetry: Tracks Event class
+ *
+ * @package Automattic\VIP\Telemetry\Tracks
+ */
+
+declare(strict_types=1);
+
+namespace Automattic\VIP\Telemetry\Tracks;
+
+use WP_Error;
+use JsonSerializable;
+use Automattic\VIP\Support_User\User as Support_User;
+use function Automattic\VIP\Logstash\log2logstash;
+
+/**
+ * Class that creates and validates Tracks events.
+ *
+ * @since 3.12.0
+ */
+class Tracks_Event implements JsonSerializable {
+	/**
+	 * Event name regex. Spaces, mixed case, and special characters are not allowed.
+	 */
+	protected const EVENT_NAME_REGEX = '/^[a-z_][a-z0-9_]*$/';
+
+	/**
+	 * Property name regex. Event props should be in snake_case. Example: compressed_size is correct, but compressedSize is not.
+	 * Property names with leading underscores are reserved for special properties.
+	 */
+	protected const PROPERTY_NAME_REGEX = '/^[a-z_][a-z0-9_]*$/';
+
+	/**
+	 * Variable containing the event's data or an error if one was encountered
+	 * during the event's creation.
+	 *
+	 * @var Tracks_Event_DTO|WP_Error
+	 */
+	protected $data = null;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param string                            $prefix The event's prefix.
+	 * @param string                            $event_name The event's name.
+	 * @param array<string, mixed>|array<empty> $event_properties Any properties included in the event.
+	 */
+	public function __construct( string $prefix, string $event_name, array $event_properties = [] ) {
+		$event_data        = static::process_properties( $prefix, $event_name, $event_properties );
+		$validation_result = static::get_event_validation_result( $event_data );
+
+		$this->data = $validation_result ?? $event_data;
+	}
+
+	/**
+	 * Returns the event's data.
+	 *
+	 * @return Tracks_Event_DTO|WP_Error Event object if the event was created successfully, WP_Error otherwise.
+	 */
+	public function get_data() {
+		return $this->data;
+	}
+
+	/**
+	 * Returns the event's data for JSON representation.
+	 */
+	#[\ReturnTypeWillChange]
+	public function jsonSerialize() {
+		if ( is_wp_error( $this->data ) ) {
+			return (object) [];
+		}
+
+		return $this->data;
+	}
+
+	/**
+	 * Returns whether the event can be recorded.
+	 *
+	 * @return bool|WP_Error True if the event is recordable.
+	 */
+	public function is_recordable() {
```

Comment:
Is there a reason why we can't encode the `bool|WP_Error` return type in PHP types here?

Reply:
No good reason other than me being defensive. Since we have good test coverage, it is a good time to add them all around which I did 👍